### ISSUE TITLE
Trainer: To keep unused columns for `compute_metrics`

### ISSUE THREAD

**Issue by sadra-barikbin** [huggingface/transformers#31570]
Hi there!

Currently, columns not used by the model are removed in `self.get_*_dataloader()` upon data loader creation, but one might want to have them in `compute_metrics` (when `include_inputs_for_metrics=True`). 

My case is fine-tuning on prompt-completion's and I use tokenizer's `token_type_ids` as a mask to compute accuracy only on the completion tokens.

To this end, the best way I've come up with is to keep that column in dataset & data loader using `remove_unused_columns=False` and then remove it in `self._prepare_inputs()` by overriding it.

Is there a better way to achieve this? Generally, isn't better to move removing unused columns logic to `self._prepare_inputs` if the logic serves only as the gatekeeper for `model(**inputs)`?

**Comment by amyeroberts** [comment]
cc @muellerzr @SunMarc

**Comment by not-lain** [comment]
@sadra-barikbin 
for the trainer API you should specify `remove_unused_columns=False` in the `TrainingArguments`
checkout https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.TrainingArguments.remove_unused_columns for more info

**Comment by sadra-barikbin** [comment]
`model(**inputs)` rejects those additional columns and raises error by only using `remove_unused_columns=False`.

**Comment by sadra-barikbin** [comment]
Minimal reproduction:

```python
from typing import Dict, List
from transformers import Trainer, TrainingArguments, AutoTokenizer, AutoModelForCausalLM
from transformers import LineByLineTextDataset, DataCollatorForLanguageModeling
from datasets import Dataset

model = AutoModelForCausalLM.from_pretrained("EleutherAI/pythia-14m")
tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-14m")
tokenizer.pad_token = tokenizer.eos_token

data = Dataset.from_dict({'prompt':['A', 'B'], 'completion': ['a', 'b']})

def tokenize(example: Dict[str, str]) -> Dict[str, List[int]]:
  return tokenizer(example['prompt'], f"{example['completion']}.", return_token_type_ids=True)

dataset = data.map(tokenize, remove_columns=['prompt', 'completion'])

args = TrainingArguments(
    output_dir="test",
    report_to='none',
    remove_unused_columns=False,
    max_steps=3,
)

trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=args,
    data_collator=DataCollatorForLanguageModeling(tokenizer, mlm=False),
    train_dataset=dataset,
)

trainer.train()
```

**Comment by zucchini-nlp** [comment]
Might be something related to the requested feature, I am adding `predict_with_generate` in Trainer which will let us compute metrics on the generated text, also can compute only on generated part (wo prompt that was fed to the model). Track in https://github.com/huggingface/transformers/pull/32346
